File: bindings/java/src/main/java/io/omq/Message.java
package io.omq;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.OptionalInt;

/** Immutable OMQ message with one or more byte-array parts. */
public final class Message {
    private final byte[] body;
    private final byte[][] parts;
    private final int routingId;

    private Message(byte[][] parts) {
        this(parts, true, 0);
    }

    private Message(byte[][] parts, boolean copy) {
        this(parts, copy, 0);
    }

    private Message(byte[][] parts, boolean copy, int routingId) {
        if (parts.length == 0) {
            throw new IllegalArgumentException("message must contain at least one part");
        }
        this.body = null;
        this.parts = copy ? copy(parts) : requireParts(parts);
        this.routingId = requireRoutingId(routingId);
    }

    private Message(byte[] body, boolean copy) {
        this(body, copy, 0);
    }

    private Message(byte[] body, boolean copy, int routingId) {
        Objects.requireNonNull(body, "body");
        this.body = copy ? Arrays.copyOf(body, body.length) : body;
        this.parts = null;
        this.routingId = requireRoutingId(routingId);
    }

    /** Creates a single-part binary message by copying {@code body}. */
    public static Message of(byte[] body) {
        Objects.requireNonNull(body, "body");
        return new Message(body, true);
    }

    /** Creates a single-part message from remaining bytes without changing position. */
    public static Message of(ByteBuffer body) {
        Objects.requireNonNull(body, "body");
        return of(readRemaining(body));
    }

    /** Creates a UTF-8 single-part text message. */
    public static Message text(String text) {
        return text(text, StandardCharsets.UTF_8);
    }

    /** Creates a single-part text message with the supplied charset. */
    public static Message text(String text, Charset charset) {
        Objects.requireNonNull(text, "text");
        Objects.requireNonNull(charset, "charset");
        return of(text.getBytes(charset));
    }

    /** Creates a multipart message by copying every part. */
    public static Message multipart(byte[]... parts) {
        Objects.requireNonNull(parts, "parts");
        return new Message(parts);
    }

    /** Creates multipart message parts from remaining bytes without changing positions. */
    public static Message multipart(ByteBuffer... parts) {
        Objects.requireNonNull(parts, "parts");
        byte[][] out = new byte[parts.length][];
        for (int i = 0; i < parts.length; i++) {
            out[i] = readRemaining(Objects.requireNonNull(parts[i], "part " + i));
        }
        return new Message(out);
    }

    static Message fromNative(byte[][] parts) {
        Objects.requireNonNull(parts, "parts");
        return new Message(parts, false);
    }

    static Message fromNative(byte[] body) {
        return new Message(body, false);
    }

    static Message fromNative(byte[] body, int routingId) {
        return new Message(body, false, routingId);
    }

    static Message fromNative(byte[][] parts, int routingId) {
        return new Message(parts, false, routingId);
    }

    static Message fromNative(Object nativeMessage) {
        if (nativeMessage instanceof Message message) {
            return message;
        }
        if (nativeMessage instanceof byte[] body) {
            return fromNative(body);
        }
        if (nativeMessage instanceof byte[][] nativeParts) {
            return fromNative(nativeParts);
        }
        throw new IllegalArgumentException("native message must be byte[] or byte[][]");
    }

    static byte[] bytesFromNative(Object nativeMessage) {
        if (nativeMessage instanceof byte[] nativeBody) {
            return nativeBody;
        }
        if (nativeMessage instanceof byte[][] nativeParts && nativeParts.length == 1) {
            return Objects.requireNonNull(nativeParts[0], "part 0");
        }
        if (nativeMessage instanceof byte[][] nativeParts) {
            throw new IllegalStateException("message has " + nativeParts.length + " parts");
        }
        throw new IllegalArgumentException("native message must be byte[] or byte[][]");
    }

    byte[][] toNative() {
        if (body != null) {
            return new byte[][] {Arrays.copyOf(body, body.length)};
        }
        return copy(parts);
    }

    /** Returns the number of message parts. */
    public int partCount() {
        return body != null ? 1 : parts.length;
    }

    /** Returns whether this message has more than one part. */
    public boolean isMultipart() {
        return body == null && parts.length > 1;
    }

    /** Returns the native routing ID, if this message came from a routed socket. */
    public OptionalInt routingId() {
        return routingId == 0 ? OptionalInt.empty() : OptionalInt.of(routingId);
    }

    /** Returns a copy of this message carrying {@code routingId} metadata. */
    public Message withRoutingId(int routingId) {
        if (body != null) {
            return new Message(body, true, routingId);
        }
        return new Message(parts, true, routingId);
    }

    /** Returns a copy of the part at {@code index}. */
    public byte[] part(int index) {
        if (body != null) {
            if (index != 0) {
                throw new ArrayIndexOutOfBoundsException(index);
            }
            return Arrays.copyOf(body, body.length);
        }
        return Arrays.copyOf(parts[index], parts[index].length);
    }

    /** Returns a read-only buffer containing a copy of the part at {@code index}. */
    public ByteBuffer partBuffer(int index) {
        return ByteBuffer.wrap(part(index)).asReadOnlyBuffer();
    }

    /** Returns copies of all message parts. */
    public List<byte[]> parts() {
        if (body != null) {
            return List.of(Arrays.copyOf(body, body.length));
        }
        return Arrays.stream(parts)
                .map(part -> Arrays.copyOf(part, part.length))
                .toList();
    }

    /** Returns the body of a single-part message. */
    public byte[] bytes() {
        if (body != null) {
            return Arrays.copyOf(body, body.length);
        }
        if (parts.length == 1) {
            return Arrays.copyOf(parts[0], parts[0].length);
        }
        throw new IllegalStateException("message has " + parts.length + " parts");
    }

    /** Decodes a single-part message as UTF-8 text. */
    public String text() {
        return text(StandardCharsets.UTF_8);
    }

    /** Decodes a single-part message as text with the supplied charset. */
    public String text(Charset charset) {
        Objects.requireNonNull(charset, "charset");
        return new String(bytes(), charset);
    }

    /** Returns whether this message has the same parts and bytes as another message. */
    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Message message) || partCount() != message.partCount()) {
            return false;
        }
        for (int i = 0; i < partCount(); i++) {
            if (!Arrays.equals(partView(i), message.partView(i))) {
                return false;
            }
        }
        return true;
    }

    /** Returns a hash code derived from message parts and bytes. */
    @Override
    public int hashCode() {
        int hash = 1;
        for (int i = 0; i < partCount(); i++) {
            hash = 31 * hash + Arrays.hashCode(partView(i));
        }
        return hash;
    }

    /** Returns a compact description without copying or exposing message bytes. */
    @Override
    public String toString() {
        return "Message[parts=" + partCount() + ", bytes=" + totalBytes() + "]";
    }

    private static byte[][] copy(byte[][] input) {
        requireParts(input);
        byte[][] out = new byte[input.length][];
        for (int i = 0; i < input.length; i++) {
            out[i] = Arrays.copyOf(input[i], input[i].length);
        }
        return out;
    }

    private static byte[][] requireParts(byte[][] input) {
        Objects.requireNonNull(input, "parts");
        for (int i = 0; i < input.length; i++) {
            Objects.requireNonNull(input[i], "part " + i);
        }
        return input;
    }

    private static byte[] readRemaining(ByteBuffer buffer) {
        ByteBuffer copy = buffer.asReadOnlyBuffer();
        byte[] out = new byte[copy.remaining()];
        copy.get(out);
        return out;
    }

    private byte[] partView(int index) {
        if (body != null) {
            if (index != 0) {
                throw new ArrayIndexOutOfBoundsException(index);
            }
            return body;
        }
        return parts[index];
    }

    private long totalBytes() {
        long total = 0;
        for (int i = 0; i < partCount(); i++) {
            total += partView(i).length;
        }
        return total;
    }

    private static int requireRoutingId(int routingId) {
        if (routingId < 0) {
            throw new IllegalArgumentException("routing ID must be non-negative");
        }
        return routingId;
    }
}

File: bindings/go/message.go
package omq

// Message is an immutable OMQ message with zero or more parts.
type Message struct {
	parts     [][]byte
	routingID uint32
}

// NewMessage copies parts into a message.
func NewMessage(parts ...[]byte) Message {
	out := Message{parts: make([][]byte, len(parts))}
	for i, part := range parts {
		out.parts[i] = append([]byte(nil), part...)
	}
	return out
}

// Bytes creates a single-part binary message.
func Bytes(data []byte) Message {
	return NewMessage(data)
}

// String creates a single-part UTF-8 message.
func String(data string) Message {
	return NewMessage([]byte(data))
}

// Multipart creates a multipart message.
func Multipart(parts ...[]byte) Message {
	return NewMessage(parts...)
}

// Group creates a RADIO/DISH group message.
func Group(group string, body []byte) Message {
	return Multipart([]byte(group), body)
}

// WithRoutingID attaches the opaque routing ID supplied by a SERVER socket.
func (m Message) WithRoutingID(routingID uint32) Message {
	m.routingID = routingID
	return m
}

// RoutingID returns the SERVER routing ID and whether it is present.
func (m Message) RoutingID() (uint32, bool) {
	return m.routingID, m.routingID != 0
}

// Route prefixes a message with a routing identity.
func Route(identity []byte, body Message) Message {
	parts := make([][]byte, 0, len(body.parts)+1)
	parts = append(parts, identity)
	parts = append(parts, body.parts...)
	return NewMessage(parts...)
}

// Parts returns copies of all message parts.
func (m Message) Parts() [][]byte {
	out := make([][]byte, len(m.parts))
	for i, part := range m.parts {
		out[i] = append([]byte(nil), part...)
	}
	return out
}

func (m Message) partsView() [][]byte {
	return m.parts
}

// Bytes returns a copy of the first part.
func (m Message) Bytes() []byte {
	part := m.Part(0)
	return part
}

// BytesOK returns the single part when the message has exactly one part.
func (m Message) BytesOK() ([]byte, bool) {
	if len(m.parts) != 1 {
		return nil, false
	}
	return append([]byte(nil), m.parts[0]...), true
}

// Part returns a copy of one part or nil when index is out of range.
func (m Message) Part(index int) []byte {
	part, ok := m.PartOK(index)
	if !ok {
		return nil
	}
	return part
}

// PartOK returns a copy of one part and whether it exists.
func (m Message) PartOK(index int) ([]byte, bool) {
	if index < 0 || index >= len(m.parts) {
		return nil, false
	}
	return append([]byte(nil), m.parts[index]...), true
}

// String decodes the first part as UTF-8 text.
func (m Message) String() string {
	if len(m.parts) == 0 {
		return ""
	}
	return string(m.parts[0])
}

// Len returns the number of parts.
func (m Message) Len() int {
	return len(m.parts)
}

// IsMultipart reports whether the message has more than one part.
func (m Message) IsMultipart() bool {
	return len(m.parts) > 1
}

// ByteLen returns the total bytes across all parts.
func (m Message) ByteLen() int {
	var total int
	for _, part := range m.parts {
		total += len(part)
	}
	return total
}

// Empty reports whether the message has no parts.
func (m Message) Empty() bool {
	return len(m.parts) == 0
}

// Route returns the first part as a routing identity.
func (m Message) Route() []byte {
	return m.Part(0)
}

// RouteOK returns the routing identity and whether it exists.
func (m Message) RouteOK() ([]byte, bool) {
	return m.PartOK(0)
}

// Group returns the first part as a group name.
func (m Message) Group() string {
	group, ok := m.GroupOK()
	if !ok {
		return ""
	}
	return group
}

// GroupOK returns the group name and whether it exists.
func (m Message) GroupOK() (string, bool) {
	part, ok := m.PartOK(0)
	if !ok {
		return "", false
	}
	return string(part), true
}

// Body returns all parts after the first part.
func (m Message) Body() Message {
	if len(m.parts) <= 1 {
		return Message{}
	}
	return NewMessage(m.parts[1:]...)
}

// Equal reports whether two messages have identical parts.
func (m Message) Equal(other Message) bool {
	if m.routingID != other.routingID || len(m.parts) != len(other.parts) {
		return false
	}
	for i, part := range m.parts {
		if string(part) != string(other.parts[i]) {
			return false
		}
	}
	return true
}
